src/qibo/tensorflow/custom_operators/cc/kernels/apply_gate_kernels.cc

```diff
@@ -262,17 +270,43 @@ struct CollapseStateFunctor<CPUDevice, T> {
       return i;
     };
 
+    // Define vector that holds norms during parallel calculation
+    int nnorms = ncores + 1;
+    if (nreps > 0 && nstates / nreps > nnorms) {
+      nnorms = nstates / nreps;
+    }
+    Eigen::VectorXf norms(nnorms);
```

Comment:
The code looks fine, but I think here we should use the tensorflow TType<T> object instead of calling directly Eigen.

Reply:
As far as I can see in [tensor_types.h](https://github.com/tensorflow/tensorflow/blob/a1bc56203f21a5a4995311825ffaba7a670d7747/tensorflow/core/framework/tensor_types.h), which I believe does the mapping between Eigen objects and TTypes, there is no TTypes equivalent to `Eigen::VectorXf`, only to `Eigen::TensorMap`. Is this the type we would like to use here? I guess that there is a way to do it with `TensorMap` but we would have to change the code slightly.

By the way, I think this works even with a `std::vector` in the place of `Eigen::VectorXf`. We just need to change `.sum()` to `std::accumulate`. I haven't done any benchmarks to see if there is any performance difference.

Reply:
Ok, ideally I believe we need a `UnalignedVec`. I am fine with VectorXf, however you are hardcoding the type, I think there is a VectorXd, right?

Reply:
One way I could make it work with `UnalignedVec` is the following:
```C++
double storage[nnorms];
typename TTypes<double>::UnalignedVec norms(storage, nnorms);
norms.setZero();
...
    norms(n) += AbsSquare(state[GetIndex(g, res)]);
...
auto norm = std::sqrt(((Eigen::Tensor<double, 0, 1, long int>)norms.sum())(0));
```
There might be a more Tensorflow-y way to convert `norms.sum()` to tensor to avoid the `Eigen::Tensor`. As far as I understand `UnalignedVec` and all other `TTypes` are `Eigen::TensorMap` and cannot allocate memory themeselves, they just point to existing objects. So I am not sure if it is possible to avoid the `storage` variable (or something equivalent).

Regarding the type, this is a good point and in fact `Eigen::VectorXd` is probably a better choice indeed. However in some sense this also hardcodes the type to `double` as does the `UnalignedVec` method above. Ideally we would like something that follows the real part of `T`, which I am not sure how to get at the type level. Another solution is to just use the complex `T` type for the norm but we would have to carry around a 0 as the imaginary part which would be suboptimal (perhaps without significant difference).

Reply:
Yes, this can be quite annoying, so I may suggest to use the double norm with VectorXd. For sure we can extract the type of the real/imag part of T, but then we need another mechanism to select the appropriate type from Eigen, or simply use C primitives with a template class.

Reply:
After some experimentation I could implement this with variable type as follows: 

The `norms` vector is declared as:
```C++
decltype(state[0].real()) norm;
decltype(norm) normstorage[nnorms];
typename TTypes<decltype(norm)>::UnalignedVec norms(normstorage, nnorms);
```
and the final norm is calculated as:
```C++
const Eigen::Tensor<decltype(norm), 0, 1, int> normsum = norms.sum();
norm = std::sqrt(normsum(0));
```
This seems to work, however it shows a strange variation in performance. For example for 30 qubits I get times ranging from 2.5 to 7 seconds, while the current version (with `VectorXf`) is stable to around 3.5 seconds.

Another strange observation is that `VectorXd` is actually the fastest. Specifically I get the following times and are pretty stable:
nqubits | VectorXd | VectorXf
-- | -- | --
28 | 0.71076| 0.85471
29 | 1.38268 | 1.91229
30 | 2.81394 | 3.68727
31 | 5.49702 | 7.67427
32 | 11.3545 | 14.4606

All methods take the same memory. Let me know which approach you think is the best.

Reply:
Thanks for the investigation. I think the decltype+typename are slowing down the process, so I fully agree in using a VectorXd instead.